rules.d/99-systemd.rules.in

```diff
@@ -18,6 +18,10 @@ SUBSYSTEM=="ubi", TAG+="systemd"
 
 SUBSYSTEM=="block", TAG+="systemd"
 
+# When a dm device is first created, it's just an empty container. Ignore it.
+# DM_NAME is not set in this case, but it's set on spurious "add" events that occur later.
+SUBSYSTEM=="block", ACTION=="add", KERNEL=="dm-*", ENV{DM_NAME}!="?*", ENV{SYSTEMD_READY}="0"
```

Comment:
hmm, what? DM fires multiple "add" events for the same DM device? uh. that's, well, broken?

devices really should show up via "add", then possibly get a series of "change" events (maybe "bind", "unbind" and so on), and eventually a "remove". But a device where the clean codepath results in multiple "add" events is really weird...

Reply:
> hmm, what? DM fires multiple "add" events for the same DM device? uh. that's, well, broken?

No, it doesn't. The other "add" events that we must handle are the synthetic/spurious ones, mostly during coldplug.

Reply:
Our general rule is that whenever appropriate we want that rules that set properties do so in all "positive" verbs the same way, and hence the typical match would be `ACTION!="remove"` rather than `ACTION=="add"`. This rule you are adding contradicts that, please change or explain.

Reply:
This is what makes dm events special: we have to act on *kernel change* events and *synthetic add* events. The purpose of the code above is to distinguish the *synthetic add* events (important) from the *kernel add" events (to be ignored).

Reply:
and why can't the synthetic uevent handling come to the same SYSTEMD_READY conclusions as the original one? why do we need to check the db for that?

Reply:
> But how does the above rule distinguish these cases, and why?

**How**: using `DM_NAME`, which is not set for the initial "add" event.

**Why**: when the kernel "add" event arrives, the device is an empty container with no table, not useful for anything. As @prajnoha wrote [above](https://github.com/systemd/systemd/pull/31661#issuecomment-1983426438), this event must be ignored.

If you want to know the details, see [10-dm.rules](https://gitlab.com/lvmteam/lvm2/-/blob/main/udev/10-dm.rules.in?ref_type=heads).

> and why can't the synthetic uevent handling come to the same SYSTEMD_READY conclusions as the original one? why do we need to check the db for that?

The two events will come to *different* conclusions.

Between the kernel "add" event and the synthetic / coldplug "add" event, there are one ore more "change" events. These can happen for various reasons. The case that matters is a "change" event caused by a dm table (re)load (more precisely, an DM device resume after changing the active table). In this event, the device properties are communicated by the kernel via `DM_COOKIE`. They need to be remembered in the udev db, because the kernel will send the cookie only once. If this change event occurs during initramfs processing, we need a coldplug "add" event to make the device known after pivot root, and we need to remember the properties for that as well (that's why dm devices need `db_persist`).

Further complication arises because the device may be inaccesible when the "reload" event (or the corresponding later coldplug event) is received. In that case, `blkid` or other stacked subsystems can't probe it. You can see in @prajnoha's example above that this is the default case for newly created LVs. The dm properties must be remembered, and the device activated on the first uevent where the dm devices are properly set up *and* the device is accessible.

@poettering, here is a question for you: would any bad happen if we did **not** set `SYSTEMD_READY=0` in such a case? The device wouldn't have any of the properties that are set by `blkid`. You created the commits that first set `SYSTEMD_READY=0` for these situations yourself back in 2011 (b7cf1b68a79833411851fa547ac9b4906207c224, 444d9432358c7ef8d1a85d330089b0027e740612). I suppose you had reasons to do so, but you were working with a dm ruleset that was different from what we have now.

If we find out that setting `SYSTEMD_READY=0` isn't necessary for incompletely initialized dm devices, we can eliminate this code from `99-systemd.rules` completely.

Reply:
I've tried hard to break a system by *not* setting `ENV{SYSTEMD_READY}="0"` in the rule above, and failed. I believe it may actually be possible to skip this rule.

But I think that needs more testing and discussion. It would be a relatively risky change to make. I'd prefer to have this PR merged as-is for now, and to discuss removal of the entire dm-specific code in `99-systemd.rules` separately, perhaps in a follow-up PR, which I would happily submit.